AbstractConfiguration.getProjectFolder() fails with IllegalStateException on a Gradle submodule with build.gradle.kts

### Description of the bug

The check at AbstractConfiguration:221 checks for the existence of `pom.xml` and `build.gradle`, however Gradle also supports Kotlin-based build files named `build.gradle.kts`; Flow should therefore check for the existence of that file too.

### Expected behavior

AbstractConfiguration.getProjectFolder() should correctly assume the folder to be a project folder if `build.gradle.kts` is present.

### Minimal reproducible example

1. git clone https://github.com/mvysny/karibu-testing/
2. Edit `gradle/libs.versions.toml` and change vaadin24 and vaadin24_spring at line 8 to 24.3.4
3. `./gradlew test`

### Versions

- Vaadin / Flow version: 24.3.4 / 24.3.5
- Java version: openjdk-21
- OS version: Linux Ubuntu 23.10

But actually, the error I get is not the same as the one in the issue description.
My error comes directly from Hilla , not from Flow `AbstractConfiguration.getProjectFolder()` method.

```
Caused by: com.vaadin.flow.server.ExecutionFailedException: Failed to determine project directory for dev mode. Directory '/home/marco/tmp/trash/xxxxxxx/gradle/vaadinNestedProjectsGradlePluginVersionMissmatch/sub-project' does not look like a Maven or Gradle project.
        at com.vaadin.hilla.internal.AbstractTaskEndpointGenerator.prepareEngineConfiguration(AbstractTaskEndpointGenerator.java:99)
        at com.vaadin.hilla.internal.AbstractTaskEndpointGenerator.getEngineConfiguration(AbstractTaskEndpointGenerator.java:72)
        at com.vaadin.hilla.internal.TaskGenerateOpenAPIImpl.execute(TaskGenerateOpenAPIImpl.java:85)
        at com.vaadin.flow.server.frontend.NodeTasks.execute(NodeTasks.java:342)
        at com.vaadin.flow.plugin.base.BuildFrontendUtil.runNodeUpdater(BuildFrontendUtil.java:332)
        at com.vaadin.gradle.VaadinBuildFrontendTask.vaadinBuildFrontend(VaadinBuildFrontendTask.kt:76)
```

Hilla also has a `GradleRunner.forProject()` methods, that checks if `build.gradle` or `build.gradle.kts` is present. If I rename the sub-project build file to `buid.gradle.kts` the build works fine.

--- a/flow-server/src/main/java/com/vaadin/flow/server/AbstractConfiguration.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/AbstractConfiguration.java
@@ -230,9 +230,9 @@ public interface AbstractConfiguration extends Serializable {
          */
         String baseDirCandidate = System.getProperty("user.dir", ".");
         Path path = Paths.get(baseDirCandidate);
-        if (path.toFile().isDirectory()
-                && (path.resolve("pom.xml").toFile().exists()
-                        || path.resolve("build.gradle").toFile().exists())) {
+        if (path.toFile().isDirectory() && (path.resolve("pom.xml").toFile()
+                .exists() || path.resolve("build.gradle").toFile().exists()
+                || path.resolve("build.gradle.kts").toFile().exists())) {
             return path.toAbsolutePath().toFile();
         } else {
             throw new IllegalStateException(String.format(

--- a/flow-server/src/test/java/com/vaadin/flow/server/AbstractConfigurationTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/AbstractConfigurationTest.java
@@ -1,0 +1,102 @@
+/*
+ * Copyright 2000-2024 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.vaadin.flow.server;
+
+import java.io.File;
+import java.io.IOException;
+
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+public class AbstractConfigurationTest {
+
+    @Rule
+    public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    AbstractConfiguration configuration = new AbstractConfiguration() {
+        @Override
+        public boolean isProductionMode() {
+            return false;
+        }
+
+        @Override
+        public String getStringProperty(String name, String defaultValue) {
+            return null;
+        }
+
+        @Override
+        public boolean getBooleanProperty(String name, boolean defaultValue) {
+            return false;
+        }
+    };
+
+    @Test
+    public void getProjectFolder_mavenProject_detected() throws IOException {
+        assertProjectFolderDetected("pom.xml");
+    }
+
+    @Test
+    public void getProjectFolder_gradleProject_detected() throws IOException {
+        assertProjectFolderDetected("build.gradle");
+    }
+
+    @Test
+    public void getProjectFolder_gradleKotlinProject_detected()
+            throws IOException {
+        assertProjectFolderDetected("build.gradle.kts");
+    }
+
+    @Test
+    public void getProjectFolder_unknownProject_throws() throws IOException {
+        withTemporaryUserDir(() -> {
+            IllegalStateException exception = Assert.assertThrows(
+                    IllegalStateException.class,
+                    configuration::getProjectFolder);
+            Assert.assertTrue(exception.getMessage().contains(
+                    "Failed to determine project directory for dev mode"));
+            Assert.assertTrue(exception.getMessage()
+                    .contains(temporaryFolder.getRoot().getAbsolutePath()));
+        });
+    }
+
+    private void assertProjectFolderDetected(String projectFile)
+            throws IOException {
+        temporaryFolder.newFile(projectFile);
+        withTemporaryUserDir(() -> {
+            File projectFolder = configuration.getProjectFolder();
+            Assert.assertEquals(temporaryFolder.getRoot(), projectFolder);
+        });
+    }
+
+    private void withTemporaryUserDir(Runnable test) throws IOException {
+        String userDir = System.getProperty("user.dir");
+        try {
+            System.setProperty("user.dir",
+                    temporaryFolder.getRoot().getAbsolutePath());
+            test.run();
+        } finally {
+            if (userDir != null) {
+                System.setProperty("user.dir", userDir);
+            } else {
+                System.clearProperty("user.dir");
+            }
+        }
+    }
+
+}

